Added metrics with accuracy threshold (#422)

neilmehta31 opened the pull request:
Added metrics with threshold as calculated from accuracy metric as specified in the issue #422

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,79 @@ def binary_classification(target, predictions, sample_weight=None):
 
         threshold = float(max_metrics["accuracy"]["threshold"])
 
+        accu_threshold_details = {
```
please change the variable name to `metrics_at_accuracy_threshold`

neilmehta31 replied:
Okay

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,79 @@ def binary_classification(target, predictions, sample_weight=None):
+            "mcc": [],
+        }
+
+        for i in range(STEPS):
```
I dont think that loop is needed here. You just want to compute response for threshold that maximize the accuracy metric.

In the code above I've used loop over steps to check different threshold values, here you have the threshold value.

neilmehta31 replied:
Yes will remove the loop and make the changes

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,79 @@ def binary_classification(target, predictions, sample_weight=None):
+            },  # there is no threshold for AUC
+            "f1": {
+                "score": np.max(accu_threshold_details["f1"]),
+                "threshold": details["threshold"][np.argmax(details["accuracy"])],
```
maybe it is worth to add variable `accuracy_threshold` instead of re-running `details["threshold"][np.argmax(details["accuracy"])]`?

neilmehta31 replied:
Alright

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -175,6 +245,9 @@ def binary_classification(target, predictions, sample_weight=None):
         return {
             "metric_details": pd.DataFrame(details),
             "max_metrics": pd.DataFrame(max_metrics),
+            "accuracy_threshold_max_metric": pd.DataFrame(
```
maybe change the name to `metrics_at_accuracy_threshold`

neilmehta31 replied:
`accuracy_threshold_max_metric` is different than `accu_threshold_details`.
Changed the later to `metrics_at_accuracy_threshold` as you told in the first comment

pplonski reviewed (changes requested):
@neilmehta31 thank you for your PR! I've added comments.

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,71 @@ def binary_classification(target, predictions, sample_weight=None):
+        metrics_at_accuracy_threshold["recall"] += [
+            recall_score(target, response, sample_weight=sample_weight)
+        ]
+        if STEPS == 0:
```
What is the purpose of this if-statement?

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,71 @@ def binary_classification(target, predictions, sample_weight=None):
 
         threshold = float(max_metrics["accuracy"]["threshold"])
```
what is the difference between `threshold` and `accuracy_threshold`?

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,71 @@ def binary_classification(target, predictions, sample_weight=None):
+            "mcc": [],
+        }
+
+        accuracy_threshold = details["threshold"][np.argmax(details["accuracy"])]
```
Is it the same as `threshold`?

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,71 @@ def binary_classification(target, predictions, sample_weight=None):
 
         threshold = float(max_metrics["accuracy"]["threshold"])
 
+        metrics_at_accuracy_threshold = {
```
do you need this variable or you can jump into line 172?

pplonski commented on supervised/utils/additional_metrics.py:
```diff
@@ -137,6 +137,71 @@ def binary_classification(target, predictions, sample_weight=None):
+            ]
+
+        # accuracy_threshold_max_metric metrics
+        accuracy_threshold_max_metric = {
```
please change the name of the variable, maybe to `accuracy_threshold_metrics`?

pplonski reviewed (changes requested):
Thank you for changes. It is very hard to read the code changes. Right now I see that you mostly copy-paste the old code without much understanding. The variable from line 140 is not needed. The if-statement from line 164 has dead-code that is never executed. Please try to refactor the code to be easy to read and understand.

neilmehta31 commented:
Hey @pplonski, Thanks for being patient. I had my evaluative components in college and hence was unable to work on the PR.
I changed the code, refactored it and made it more readable and easy to understand. I will put more emphasis on code quality in my future contributions. Please let me know if some more things need to be changed.
Thanks
